### PR TITLE
Make default output less noisy

### LINKED ISSUE #74: Make the default output less noisy
The default output listing every file to delete is quite noisy, since it prints every file that will be deleted.

We should by default no longer list each file being deleted for toolchain, just list the project path instead and the total size removed.

_Originally posted by @juhp in https://github.com/holmgr/cargo-sweep/pull/60#issue-1014328857

### PR DESCRIPTION
**Pull request by marcospb19**
Fixes #74.

`cargo-sweep` logs every deleted file, this makes the output
very noisy and hard to follow (outputs approximately a thousand
lines per cleaned project).

This PR changes the default output to omit deleted files, and
only show them with the `--verbose` flag.

### CODE REVIEW

**Comment by marcospb19** [comment]
PR is a draft because this is the remaining output after the changes:
```
[INFO] Cleaned 5.34 GiB
[INFO] Using all installed toolchains: ["stable-2022-09-22-x86_64-unknown-linux-
gnu", "stable-x86_64-unknown-linux-gnu", "nightly-2022-06-20-x86_64-unknown-linu
x-gnu", "nightly-2022-08-06-x86_64-unknown-linux-gnu", "nightly-x86_64-unknown-l
inux-gnu", "1.63.0-x86_64-unknown-linux-gnu", "1.64.0-x86_64-unknown-linux-gnu"]
```

So now you don't know what project was cleaned.

**Comment by marcospb19** [comment]
The list of installed toolchains is shown for every project it sweeps, but I think it is always the same.

So we should probably just show it once at the start.

This is the function that loads the toolchains:

https://github.com/holmgr/cargo-sweep/blob/75414ca5dfd138691031c6f701fbaa350fd5b095/src/fingerprint.rs#L373-L374

It's always constant, because it calls the `rustup` binary, I think we can assume the binary doesn't change (also, cargo sweep nevers `cd` into other directory).

**Comment by marcospb19** [comment]
Should be ready for review now, fixing the list of toolchains being outputted repeatedly will require a little bit of refactoring, so I prefer doing it in another PR.

**Comment by marcospb19** [comment]
Sorry I had to force push again, `cargo clippy --fix` broke a test.

File: tests/integration.rs [hunk 1]
```diff
@@ -49,6 +49,7 @@ fn sweep(args: &[&str]) -> Command {
     let mut cmd = Command::new(cargo_bin("cargo-sweep"));
     cmd.arg("sweep")
         .current_dir(project_dir("sample-project"))
+        .arg("--verbose")
```
Comment (marcospb19) [inline tests/integration.rs:52]: I added --verbose to all tests here, and removed it from the only test that required it (stamp_file function), because specifying this flag twice makes clap complain.

**Comment by jyn514** [comment]
I just fixed the clippy warnings in #80 (and made a couple other changes that should make the errors easier to read) - do you mind rebasing over master?

Also, please revert the changes to the assertion message, like the error says they don't do what you expect in 2018 edition and I would prefer to switch the edition in a separate PR.

**Comment by marcospb19** [comment]
Oh ok, I'll do that, I was having a lot of trouble with the assertion message, so I force pushed like 5 times, let me fix it.

File: tests/integration.rs [hunk 2]
```diff
@@ -102,6 +111,7 @@ fn clean_and_parse(target: &TempDir, args: &[&str]) -> Result<u64> {
     Ok(cleaned)
 }
 
+#[allow(clippy::uninlined_format_args)]
```
Comment (marcospb19) [inline tests/integration.rs:114]: @jyn514 actually, I added this because the _git hooks_ wouldn't let me push without it, have you solved this in the last PR?

**Comment by jyn514** [comment]
Do you mind adding a test showing the new output? Comparing `run(sweep(&["--maxsize", "0"])).get_output().stdout` to some string would be ideal; I guess it's annoying because it hashes in the name but maybe you can normalize those?

File: tests/integration.rs (hunk 2 above)
Comment (jyn514) [reply tests/integration.rs:114]: ah gotcha - you can use `--no-verify` to bypass the git hook, or rebase over master where I've fixed it.

I wonder if I should get rid of clippy in the pre-push hook and only run rustfmt instead :thinking:

File: tests/integration.rs [hunk 3]
```diff
@@ -173,23 +184,59 @@ fn stamp_file() -> TestResult {
...
+    ),
+]
+[DEBUG] cleaning: "/tmp/.tmpYzB6xH/debug" with remove_not_built_with_in_a_profile
+[DEBUG] Successfully removed: "/tmp/.tmpYzB6xH/debug/deps/libsample_project-261b05c2354104eb.rlib"
+[DEBUG] Successfully removed: "/tmp/.tmpYzB6xH/debug/deps/libsample_project-261b05c2354104eb.rmeta"
+[DEBUG] Successfully removed: "/tmp/.tmpYzB6xH/debug/deps/sample_project-261b05c2354104eb.d"
+[DEBUG] Successfully removed: "/tmp/.tmpYzB6xH/debug/.fingerprint/sample-project-261b05c2354104eb"
+[INFO] Cleaned 9.43 KiB from "/tmp/.tmpYzB6xH"
+               "#
+    );
```
Comment (marcospb19) [inline tests/integration.rs:225]: @jyn514 I inserted it here just to show you, what do you think of matching with RegEx to skip all hashed values?

Comment (jyn514) [reply tests/integration.rs:225]: Thanks! That sounds like a good idea :)

File: tests/integration.rs [hunk 4]
```diff
@@ -173,23 +191,63 @@ fn stamp_file() -> TestResult {
...
+                "{regex_skip}",
+            \),
+        \]
+        \[DEBUG\] cleaning: "/tmp/{regex_skip}/debug" with remove_not_built_with_in_a_profile
+        \[DEBUG\] Successfully removed: "/tmp/{regex_skip}/debug/deps/libsample_project-{regex_skip}.rlib"
+        \[DEBUG\] Successfully removed: "/tmp/{regex_skip}/debug/deps/libsample_project-{regex_skip}.rmeta"
+        \[DEBUG\] Successfully removed: "/tmp/{regex_skip}/debug/deps/sample_project-{regex_skip}.d"
+        \[DEBUG\] Successfully removed: "/tmp/{regex_skip}/debug/.fingerprint/sample-project-{regex_skip}"
+        \[INFO\] Cleaned {regex_skip} from "/tmp/{regex_skip}""#,
+    ));
```
Comment (marcospb19) [inline tests/integration.rs:231]: This does work, but if the test fails, it doesn't give any tips on what changed, I'm not sure if it's a great approach.

Comment (marcospb19) [reply tests/integration.rs:231]: For reference, here's the output:
```
[DEBUG] cleaning: "/tmp/.tmpXhoLjT" with remove_older_until_fits
[DEBUG] size_to_remove: 46978
[DEBUG] Sizing: "/tmp/.tmpXhoLjT/debug" with total_disk_space_in_a_profile
[DEBUG] Hashs by time: [
    (
        13.926894ms,
        "261b05c2354104eb",
    ),
]
[DEBUG] cleaning: "/tmp/.tmpXhoLjT/debug" with remove_not_built_with_in_a_profile
[DEBUG] Successfully removed: "/tmp/.tmpXhoLjT/debug/deps/libsample_project-261b05c2354104eb.rlib"
[DEBUG] Successfully removed: "/tmp/.tmpXhoLjT/debug/deps/libsample_project-261b05c2354104eb.rmeta"
[DEBUG] Successfully removed: "/tmp/.tmpXhoLjT/debug/deps/sample_project-261b05c2354104eb.d"
[DEBUG] Successfully removed: "/tmp/.tmpXhoLjT/debug/.fingerprint/sample-project-261b05c2354104eb"
[INFO] Cleaned 9.43 KiB from "/tmp/.tmpXhoLjT"
```

Comment (jyn514) [reply tests/integration.rs:231]: I think it's ok; if you run the test with `--nocapture` it will show the original output.

File: tests/integration.rs [hunk 5]
```diff
@@ -173,23 +191,63 @@ fn stamp_file() -> TestResult {
...
+    let (_size, target) = build("sample-project")?;
+
+    let assert = run(
+        sweep(&["--maxsize", "0"]).current_dir(test_dir().join("sample-project")),
+        target.path(),
+    );
+
+    let output = std::str::from_utf8(&assert.get_output().stdout).unwrap();
+
+    let regex_skip = r#".+?"#;
```
Comment (jyn514) [inline tests/integration.rs:213]: this can just be `.*` instead of needing to make it optional; and at that point I think it would be simpler to inline it instead of using `format`.

**Review by jyn514 — APPROVED** [review]
Thanks! Happy to merge this with the last nit fixed :)

File: tests/integration.rs [hunk 6]
```diff
@@ -173,23 +191,63 @@ fn stamp_file() -> TestResult {
...
+
+    let pattern = unindent(&format!(
+        r#"\[DEBUG\] cleaning: "/tmp/{regex_skip}" with remove_older_until_fits
+        \[DEBUG\] size_to_remove: {regex_skip}
+        \[DEBUG\] Sizing: "/tmp/{regex_skip}/debug" with total_disk_space_in_a_profile
```
Comment (jyn514) [inline tests/integration.rs:218]: ah, and like CI shows this needs to be `".*/debug"` instead of assuming it's in /tmp.

File: tests/integration.rs (hunk 5 above)
Comment (marcospb19) [reply tests/integration.rs:213]: Aha! A regex _Gotcha_.

Fun fact: Inserting an `?` after a quantifier (`*`, `+`, `?` or `{N..M}`) switches
to non-greedy matching, which can be good for performance as it decreases
excessive backtracking from the usual greedy eater.

(It's the same symbol to make stuff optional... why...)

Well, that's a tiny text to match so I'll switch to `.+` anyways :+1:.

Comment (marcospb19) [reply tests/integration.rs]: I do think this is a bit less readable from the editor, but the GitHub diff doesn't highlight the format placeholder.

image

image

**Comment by marcospb19** [comment]
Now it's failing because of Windows paths 
```
"C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\.tmpBWHDFH\\debug\\.fingerprint\\
```

**Comment by marcospb19** [comment]
Fixed, if you want I can squash the last (one or two) commits.

**Comment by jyn514** [comment]
Thanks for sticking with this!
